### PR TITLE
Always use RGBA for DXT1 textures.

### PR DESCRIPTION
**Pull request by Armada651**
DirectX doesn't define an RGB version of DXT1, the RGB format of DXT1 is specific to OpenGL. Since DXT1 is a DirectX-specific format, I think we should follow the [specification on MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/bb147243%28v=vs.85%29.aspx).

Fixes issue #68.

### CODE REVIEW

File: gli/core/gl.inl [hunk 1]
```diff
@@ -146,8 +146,6 @@ namespace gli
 			{INTERNAL_D24S8, EXTERNAL_DEPTH_STENCIL, TYPE_NONE},					//FORMAT_D24_UNORM_S8_UINT,
 			{INTERNAL_D32FS8X24, EXTERNAL_DEPTH_STENCIL, TYPE_NONE},				//FORMAT_D32_SFLOAT_S8_UINT,
 
-			{INTERNAL_RGB_DXT1, EXTERNAL_NONE, TYPE_NONE},							//FORMAT_RGB_DXT1_UNORM,
```
Comment (Groovounet) [inline gli/core/gl.inl]: I don't think that's the right fix. OpenGL does have different enum values for DXT1 RGB and RGBA.

Comment (Armada651) [inline gli/core/gl.inl]: OpenGL does, but DirectX doesn't. That's the problem, there is nothing in
the header that specifies an alpha channel.

**Comment by Armada651** [comment]
@Groovounet Everywhere I look at the MSDN documentation it becomes increasingly clear that DXT1 with single-bit alpha has superseded the variant without single-bit alpha. Everywhere DXT1 is mentioned it assumes single-bit alpha is available:

> If 64-bit blocks - that is, format DXT1 - are used for the texture, it is possible to mix the opaque and 1-bit alpha formats on a per-block basis within the same texture.

Source: https://msdn.microsoft.com/en-us/library/windows/desktop/bb204843%28v=vs.85%29.aspx

**Comment by Groovounet** [comment]
I'll investigate but in any case I think we need to be able to express the
OpenGL enums.

My recollection of the difference between alpha or not is that there are 4
values (2 bits) per pixels to define the interpolation between color A and
B in a DXT block. When using RGB the four values are used to interrelation
between A and B. In RGBA,there are only 3 values for interpolation and 1
value for transparent pixels.

**Comment by Armada651** [comment]
@Groovounet That is not entirely correct. What you are explaining is the two different interpolation modes supported by DXT1. Blocks that define `color0 > color1` will use 4 values (2 bits) per pixels. And blocks that define `color0 <= color1` use 3 values for interpolation and 1 value for Black or Transparent, depending on whether the internal format uses RGB or RGBA.

Note that Black is only defined for OpenGL. DirectX always uses Transparent and thus always assumes the RGBA internal format.

**Comment by Armada651** [comment]
@Groovounet Have you come to a conclusion? I'd be happy to provide more information on this issue if you're still in doubt.

**Comment by Groovounet** [comment]
I didn't yet look at the image but my conclusion so far is that DXT1 RGB and DXT1 RGBA is a thing even for Direct3D because the blocks are different. So I think there is something in the header which is going to show the difference allowing me to fix this bug correctly.

**Comment by Armada651** [comment]
@Groovounet The blocks are different based on the `color0 > color1` condition, not based on the internal format. There can't be a header flag because whether the texture has 1-bit alpha is decided *per block*, so it can never be part of the header.

**Comment by Armada651** [comment]
If you still don't believe that the only difference between RGB and RGBA is whether the alpha bit expresses Black or Transparency, then please check my sources:

https://www.opengl.org/wiki/S3_Texture_Compression#DXT1_Format
https://msdn.microsoft.com/en-us/library/windows/desktop/bb147243%28v=vs.85%29.aspx

Both the OpenGL wiki and MSDN confirm my findings and furthermore MSDN makes no mention anywhere of a mode that uses the alpha bit for Black instead of Transparent. I've also tested this PR in a game which makes use of lots of DXT1 textures with no issues whatsoever.

**Comment by Groovounet** [comment]
"The interpretation differs slightly. You always get an alpha value of 1 unless the pixel uses the code for Black in the above table. In that case, you get an alpha value of 0."

So right, my recollection wasn't correct but there is still a difference between RGBA_DXT1 and RGB_DXT1.

Using TheCompressonator and stepping in the code, the headers differ as well.
RGB_DXT1 => DDPF_FOURCC
RGBA_DXT1 => DDPF_FOURCC | DDPF_ALPHAPIXELS

**Comment by Armada651** [comment]
@Groovounet But the DirectX specs make no mention anywhere of a DXT1 format where the alpha bit represents Black. You're willing to forgo DirectX compatibility because OpenGL specifies another format?

None of the DXT1 textures with alpha blocks that I have specify the DDPF_ALPHAPIXELS flag and according to the DirectX specs it should be valid even without that flag.

**Comment by Groovounet** [comment]
How this is jeopardizing DX compatibility?

And DX do the same:
"
The difference between DXT4 and DXT5 is that in the DXT4 format it is assumed that the color data has been premultiplied by alpha. In the DXT5 format, it is assumed the color is not premultiplied by alpha. These two formats are needed because, in most cases, by the time a texture is used, just examining the data is not sufficient to determine if the color values have been multiplied by alpha. Because this information is needed at run time, the two FOURCC codes are used to differentiate these cases. However, the data and interpolation method used for these two formats is identical.
The color compare used in DXT1 to determine if the texel is transparent is not used with these formats. It is assumed that without the color compare the color data is always treated as if in 4-color mode. In other words, the if statement at the top of the DXT1 code should be:
"

**Comment by Armada651** [comment]
I could send you a screenshot of an engine with both a DirectX and an OpenGL backend loading the exact same texture where DirectX will show a transparent texture and OpenGL will show a black square, would that convince you that DX compatibility is jeopardized?

**Comment by Groovounet** [comment]
No

**Comment by Armada651** [comment]
Then how can I prove it?

**Comment by Armada651** [comment]
Also note that your quote is referring to the DXT4 and DXT5 formats.

**Comment by Armada651** [comment]
And even if you would assume this where true for the DXT1 format, it would still not be correct as OpenGL specifies the color compare is still used even if the internal format is RGB.

**Pull request closed without merging**